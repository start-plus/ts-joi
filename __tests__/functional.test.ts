import {
  required,
  minLength,
  maxLength,
  optional,
  getSchema,
  integer,
  items,
  xor,
} from '../src';

it('string type', () => {
  class MyType {
    @required()
    @minLength(2)
    @maxLength(10)
    foo: string;

    @optional() bar: string;
  }
  expect(getSchema(MyType).describe()).toMatchSnapshot();
});

it('number type', () => {
  class MyType {
    @required()
    @integer()
    foo: number;
  }
  expect(getSchema(MyType).describe()).toMatchSnapshot();
});

it('array type', () => {
  class MyType {
    @items(Number)
    @required()
    foo: number[];
  }
  expect(getSchema(MyType).describe()).toMatchSnapshot();
});

type numberOrString = number | string;

it('array type (union type)', () => {
  class MyType {
    @items(Number, String)
    @required()
    foo: numberOrString[];
  }
  expect(getSchema(MyType).describe()).toMatchSnapshot();
});

it('array type (custom type)', () => {
  class Item {
    @required() id: string;
    @required() name: string;
  }
  class MyType {
    @items(Item)
    @required()
    foo: Item[];
  }
  expect(getSchema(MyType).describe()).toMatchSnapshot();
});

it('object type', () => {
  @xor('foo', 'bar')
  class MyType {
    @optional() foo?: string;

    @optional() bar?: string;
    @required() baz: string;
  }
  expect(getSchema(MyType).describe()).toMatchSnapshot();
});

it('object type (inner) ', () => {
  class Options {
    @optional() foo?: string;
    @required() bar: string;
  }

  class MyType {
    @optional() options?: Options;
  }
  expect(getSchema(MyType).describe()).toMatchSnapshot();
});

it('object type (inner with override) ', () => {
  class Options {
    @optional() foo?: string;
    @required() bar: string;
  }

  class MyType {
    @xor('foo', 'bar')
    @optional()
    options?: Options;
  }
  expect(getSchema(MyType).describe()).toMatchSnapshot();
});
