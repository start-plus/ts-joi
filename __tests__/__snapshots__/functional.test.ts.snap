// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array type (custom type) 1`] = `
Object {
  "children": Object {
    "foo": Object {
      "flags": Object {
        "presence": "required",
        "sparse": false,
      },
      "items": Array [
        Object {
          "children": Object {
            "id": Object {
              "flags": Object {
                "presence": "required",
              },
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "name": Object {
              "flags": Object {
                "presence": "required",
              },
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
          },
          "type": "object",
        },
      ],
      "type": "array",
    },
  },
  "type": "object",
}
`;

exports[`array type (union type) 1`] = `
Object {
  "children": Object {
    "foo": Object {
      "flags": Object {
        "presence": "required",
        "sparse": false,
      },
      "items": Array [
        Object {
          "invalids": Array [
            Infinity,
            -Infinity,
          ],
          "type": "number",
        },
        Object {
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
      ],
      "type": "array",
    },
  },
  "type": "object",
}
`;

exports[`array type 1`] = `
Object {
  "children": Object {
    "foo": Object {
      "flags": Object {
        "presence": "required",
        "sparse": false,
      },
      "items": Array [
        Object {
          "invalids": Array [
            Infinity,
            -Infinity,
          ],
          "type": "number",
        },
      ],
      "type": "array",
    },
  },
  "type": "object",
}
`;

exports[`number type 1`] = `
Object {
  "children": Object {
    "foo": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        Infinity,
        -Infinity,
      ],
      "rules": Array [
        Object {
          "name": "integer",
        },
      ],
      "type": "number",
    },
  },
  "type": "object",
}
`;

exports[`object type (inner with override)  1`] = `
Object {
  "children": Object {
    "options": Object {
      "children": Object {
        "bar": Object {
          "flags": Object {
            "presence": "required",
          },
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
        "foo": Object {
          "flags": Object {
            "presence": "optional",
          },
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
      },
      "dependencies": Array [
        Object {
          "key": null,
          "peers": Array [
            "foo",
            "bar",
          ],
          "type": "xor",
        },
      ],
      "flags": Object {
        "presence": "optional",
      },
      "type": "object",
    },
  },
  "type": "object",
}
`;

exports[`object type (inner)  1`] = `
Object {
  "children": Object {
    "options": Object {
      "children": Object {
        "bar": Object {
          "flags": Object {
            "presence": "required",
          },
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
        "foo": Object {
          "flags": Object {
            "presence": "optional",
          },
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
      },
      "flags": Object {
        "presence": "optional",
      },
      "type": "object",
    },
  },
  "type": "object",
}
`;

exports[`object type 1`] = `
Object {
  "children": Object {
    "bar": Object {
      "flags": Object {
        "presence": "optional",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "baz": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "foo": Object {
      "flags": Object {
        "presence": "optional",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "dependencies": Array [
    Object {
      "key": null,
      "peers": Array [
        "foo",
        "bar",
      ],
      "type": "xor",
    },
  ],
  "type": "object",
}
`;

exports[`string type 1`] = `
Object {
  "children": Object {
    "bar": Object {
      "flags": Object {
        "presence": "optional",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "foo": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "rules": Array [
        Object {
          "arg": 10,
          "name": "max",
        },
        Object {
          "arg": 2,
          "name": "min",
        },
      ],
      "type": "string",
    },
  },
  "type": "object",
}
`;
